### PR TITLE
Feature 52 - searching tags

### PR DESCRIPTION
**Pull request by jerushachua**
This is a feature implementation for issue #52 . 

I added a function which pulls the input from the search bar and displays only the matching input. The input is case sensitive, since Linux file systems are also case sensitive, but this can be easily changed if needed. 

Please feel free to help test any edge cases I may have missed.

### CODE REVIEW

File: app/misc/git.ts [hunk 1]
```diff
@@ -495,6 +495,56 @@ function createBranch() {
...
+                document.getElementById(refName).setAttribute("style","display:block");
+              }
+            }
+
+            lastRefList = refList.slice(); // update lastRefList
+        })
+    });
+}
+
+function displayTag(name, id, onclick) {
```
Comment (sterlinglaird) [inline app/misc/git.ts]: This function already exists in repo.ts?

File: app/misc/git.ts [hunk 2]
```diff
@@ -495,6 +495,56 @@ function createBranch() {
...
+function searchTag() {
+  Git.Repository.open(repoFullPath)
+    .then(function (repo) {
+      repo.getCurrentBranch()
+        .then(function () {
+          // grab the list of references - these could be branches or tags
+          return repo.getReferences(Git.Reference.TYPE.LISTALL);
+        }).then(function (refList) {
+          refList.sort();
+          bname = {};
```
Comment (sterlinglaird) [inline app/misc/git.ts]: Why are these objects getting cleared?

Comment (sterlinglaird) [reply app/misc/git.ts]: And couldnt you just get the tag name from the tags object? that way we dont need to continue querying node git

File: app/misc/git.ts (hunk 1 above)
Comment (jerushachua) [reply app/misc/git.ts]: Yes - I thought repo.ts is only called when the repo is initialized. If this isn't the case, then I can just remove the duplicate function.

**Comment by sterlinglaird** [comment]
This isn't a requirement for completion of this issue but it wouldn't be very difficult to do:
The search feature would probably be a lot more usable if every time the search bar had input all the tags which have a matching prefix show up. That way if you type in "v1." you'd see "v1.0, v1.0.1, v1.2, v1.3 etc". The "ok" button could also get removed since that is only for creating a branch, which James doesn't actually want to happen through that menu anyways.

File: app/misc/git.ts [hunk 3]
```diff
@@ -495,6 +495,56 @@ function createBranch() {
...
+          tags = {};
+          for (let i = 0; i < refList.length; i++) {
```
Comment (sterlinglaird) [inline app/misc/git.ts]: Theres a missing closing curly bracket for this for loop. My typescript compiler didn't like it

File: app/misc/git.ts (hunk 2 above)
Comment (jerushachua) [reply app/misc/git.ts]: I wasn't sure if the branches and tags would be reloaded in real time. This only queries nodegit if the tags are being searched for. It's just to ensure we have the latest copy of tags/branches. But you're right that it would be more efficient to use the tags object, if it's up to date. I can test it out with the tags object and if it's working well, I submit a changeset.

Comment (sterlinglaird) [reply app/misc/git.ts]: It probably doesn't matter too much, the object should be up to date though. Regardless, they shouldn't be cleared from within this function, thats probably more important.

**Comment by SiRumCz** [comment]
Based on my experience on old(like ancient) AnuglarJS, I think they have built-in module or mechanics to be bable to filter the tag or branch name by typing partial names. Hope it could help you.

File: app/misc/git.ts (hunk 3 above)
Comment (jerushachua) [reply app/misc/git.ts]: should be fixed now!

**Comment by jerushachua** [comment]
Fixed all the issues that were mentioned. I removed the OK button and searching is now responsive to user input. Let me know if there are any issues that I missed in my own testing.

File: app/components/header/header.component.ts [hunk 4]
```diff
@@ -1,98 +1,98 @@
...
-  showBranchList(): void {
-    this.branchSelectedInRefDropdown = true
-  }
-
-  showTagList(): void {
-    this.branchSelectedInRefDropdown = false
-  }
-
-}
+import { Component } from "@angular/core";
```
Comment (colinwerner) [inline app/components/header/header.component.ts]: What's with all the whitespace differences?

File: app/misc/git.ts [hunk 5]
```diff
@@ -495,6 +495,37 @@ function createBranch() {
...
+      repo.getCurrentBranch()
+        .then(function () {
+          // grab the list of references - these could be branches or tags
+          return repo.getReferences(Git.Reference.TYPE.LISTALL);
+        }).then(function (refList) {
+          refList.sort();
+          for (let i = 0; i < refList.length; i++) {
+
+            // strip name for readability
+            let refName = refList[i].name().split("/")[refList[i].name().split("/").length - 1];
```
Comment (colinwerner) [inline app/misc/git.ts]: Logically you can move this into the if statement?

Also, can you state what format the name is, please?
ie
name is "/blah/blah/blah/name" and we want to extract "name"?

File: app/misc/git.ts [hunk 6]
```diff
@@ -495,6 +495,37 @@ function createBranch() {
...
+
+            if (refList[i].isTag()){
+              if (refName.indexOf( document.getElementById("tag-name").value ) > -1) {
```
Comment (colinwerner) [inline app/misc/git.ts]: Just to be sure, if you find the refname in the tagname then it returns > -1?

File: app/misc/git.ts [hunk 7]
```diff
@@ -495,6 +495,37 @@ function createBranch() {
...
+                document.getElementById(refName).setAttribute("style","display:block");
+              } else {
+                document.getElementById(refName).setAttribute("style","display:none");
```
Comment (colinwerner) [inline app/misc/git.ts]: lines 515 and 517 could be refactored?
if > -1
  attr = "display:block"
else
  attr = "display:none"
document.getelementbyid(refname).setattribute("style", attr)

File: app/misc/git.ts [hunk 8]
```diff
@@ -810,12 +841,12 @@ function displayModifiedFiles() {
         if (modifiedFiles.length !== 0) {
           if (document.getElementById("modified-files-message") !== null) {
             let filePanelMessage = document.getElementById("modified-files-message");
-            filePanelMessage.parentNode.removeChild(filePanelMessage);
+            filePanelMessage.parentNode.Child(filePanelMessage);
```
Comment (colinwerner) [inline app/misc/git.ts]: I'm just curious about this, what necessitates this change?

File: app/misc/git.ts [hunk 9]
```diff
@@ -495,6 +495,37 @@ function createBranch() {
...
+// search for tags
+function searchTag() {
+  Git.Repository.open(repoFullPath)
+    .then(function (repo) {
+      repo.getCurrentBranch()
+        .then(function () {
+          // grab the list of references - these could be branches or tags
+          return repo.getReferences(Git.Reference.TYPE.LISTALL);
+        }).then(function (refList) {
+          refList.sort();
```
Comment (colinwerner) [inline app/misc/git.ts]: Is the sort necessary if we're doing an indexed search on the name?

File: app/misc/git.ts [hunk 10]
```diff
@@ -495,6 +495,37 @@ function createBranch() {
...
+            let refName = refList[i].name().split("/")[refList[i].name().split("/").length - 1];
+
+            if (refList[i].isTag()){
+              if (refName.indexOf( document.getElementById("tag-name").value ) > -1) {
+                document.getElementById(refName).setAttribute("style","display:block");
+              } else {
+                document.getElementById(refName).setAttribute("style","display:none");
+              }
+            }
+            lastRefList = refList.slice(); // update lastRefList
```
Comment (colinwerner) [inline app/misc/git.ts]: Does slice just return a copy of the array? What exactly is lastRefList used for?

File: app/misc/git.ts (hunk 5 above)
Comment (jerushachua) [reply app/misc/git.ts]: refName is the readable, string format of the tag name. The string returned from nodegit needs to be stripped before it can be used. 

Which if statement are you referring to?

File: app/misc/git.ts (hunk 6 above)
Comment (jerushachua) [reply app/misc/git.ts]: Yes, js has a built-in function which helps you compare strings. This will check if refName is a substring of the tag we're currently comparing to.

File: app/components/header/header.component.ts (hunk 4 above)
Comment (jerushachua) [reply app/components/header/header.component.ts]: I don't know why git detects this as a diff.

File: app/misc/git.ts (hunk 8 above)
Comment (jerushachua) [reply app/misc/git.ts]: I changed this to test during the functionality of removing nodes. There wasn't any effect by the change here, but I forgot to change it back. I'll change it right now and update the files.

File: app/misc/git.ts (hunk 10 above)
Comment (jerushachua) [reply app/misc/git.ts]: This was used in the original cleaning of the array when we first grab the tags. I can remove it as it's not needed here.

File: app/misc/git.ts (hunk 7 above)
Comment (jerushachua) [reply app/misc/git.ts]: Sure, done.

File: app/misc/git.ts (hunk 5 above)
Comment (colinwerner) [reply app/misc/git.ts]: It's an optimization. You are doing the parsing of the name BEFORE checking if it is a tag. You should check if it is a tag before parsing the name. Hence move let refName =... into the if (refList[i].isTag()) {

File: app/components/header/header.component.ts (hunk 4 above)
Comment (colinwerner) [reply app/components/header/header.component.ts]: Just revert it.
